File: 0001-0500/0004/0004.median-of-two-sorted-arrays.cs
/*
 * @lc app=leetcode id=4 lang=csharp
 *
 * [4] Median of Two Sorted Arrays
 */

// @lc code=start
public class Solution {
    public double FindMedianSortedArrays(int[] nums1, int[] nums2) {
        int[] merged = new int[nums1.Length + nums2.Length];
        int i = 0, j = 0, k = 0;
        while(i < nums1.Length && j < nums2.Length) {
            if(nums1[i] < nums2[j]) {
                merged[k++] = nums1[i++];
            } else {
                merged[k++] = nums2[j++];
            }
        }
        while(i < nums1.Length) {
            merged[k++] = nums1[i++];
        }
        while(j < nums2.Length) {
            merged[k++] = nums2[j++];
        }
        if(merged.Length % 2 == 0) {
            return (merged[merged.Length / 2 - 1] + merged[merged.Length / 2]) / 2.0;
        } else {
            return merged[merged.Length / 2];
        }
    }
}
// @lc code=end

File: 0001-0500/0004/0004.median-of-two-sorted-arrays.cpp
/*
 * @lc app=leetcode id=4 lang=cpp
 *
 * [4] Median of Two Sorted Arrays
 */

// @lc code=start
class Solution
{
public:
    double findMedianSortedArrays(vector<int> &nums1, vector<int> &nums2)
    {
        int total = nums1.size() + nums2.size(), mid = total / 2;
        int i = 0, j = 0, k = 0, prev = 0, curr = 0;
        while (k <= mid)
        {
            prev = curr;
            if (i < nums1.size() && j < nums2.size())
            {
                if (nums1[i] < nums2[j])
                    curr = nums1[i++];
                else
                    curr = nums2[j++];
            }
            else if (i < nums1.size())
            {
                curr = nums1[i++];
            }
            else
            {
                curr = nums2[j++];
            }
            k++;
        }
        if (total % 2 == 0)
            return (prev + curr) / 2.0;
        else
            return curr;
    }
};
// @lc code=end
